android/src/main/java/com/reactnativecommunity/webview/RNCWebView.java

```diff
@@ -242,12 +247,31 @@ public boolean getMessagingEnabled() {
         return this.messagingEnabled;
     }
 
-    protected RNCWebViewBridge createRNCWebViewBridge(RNCWebView webView) {
-        if (bridge == null) {
-            bridge = new RNCWebViewBridge(webView);
-            addJavascriptInterface(bridge, JAVASCRIPT_INTERFACE);
+    protected WebViewCompat.WebMessageListener bridgeListener = null;
+
+    protected void createRNCWebViewBridge(RNCWebView webView) {
+        if (this.bridgeListener == null &&
+                WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER)) {
+            this.bridgeListener = new WebViewCompat.WebMessageListener() {
+                @Override
+                public void onPostMessage(@NonNull WebView view, @NonNull WebMessageCompat message, @NonNull Uri sourceOrigin, boolean isMainFrame, @NonNull JavaScriptReplyProxy replyProxy) {
+                    RNCWebView.this.onMessage(message.getData(), sourceOrigin.toString());
+                }
+            };
+            WebViewCompat.addWebMessageListener(
+                    webView,
+                    JAVASCRIPT_INTERFACE,
+                    Set.of("*"),
```

Comment:
having a question here. if we put `*` here, what's the difference between original `addJavascriptInterface` approach?
maybe we can reuse the [`originWhitelist`](https://github.com/react-native-webview/react-native-webview/blob/4733fe7988f1a4489f0d430ffcbe152416fe1b0e/docs/Reference.md#originwhitelist) prop?

Reply:
This PR does not solve the fact that messages are coming from unwanted origins, i.e. origins not on a whitelist, but rather messages are coming from origin A but this library says that its coming from Origin B. 

The `*` here is just saying any origin can send messages over the bridge. The difference between the previous implementation is how we retrieve the origin of the message.

In the previous implementation we retrieve the origin of the message using `webView.getUrl()` **after** the message has been sent. The webview could be on a different page by the time we call `webView.getUrl()`.

By using a `WebMessageListener` in this PR, we get the origin **with** the message. Therefore, even if the site navigates away the instant after sending the message - the origin will still be the origin of the original sender.

Reply:
thanks for the great explanation. so to achieve the spoofing prevention as you described in #3439, the callsite (app code) have to further check the `url` from the `onMessage` callback? this pr just to make sure the url passing into onMessage is correct. is my understanding right?

Reply:
Yup! exactly. 
This PR just makes sure that the `url` on the `onMessage` is correct. 
Each app can then decide in its own way - whether they want to process or ignore the messages.

Reply:
cool! thanks for confirming that. i think the pr title may be too strong. i was scared by the title. 😅 that's not a real vulnerability, or at least we still need a further check to prevent the vulnerability.
how about changing the title like `fix(android,ios): Sending correct url in onMessage callback`

Reply:
Updated the title!